org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/LinkedHashSetQueue.java

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2018 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.builder.impl;
+
+import java.util.AbstractQueue;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Queue;
+
+/**
+ * {@link Queue} with unique valuess
+ * 
+ * @author Christian Dietrich - Initial contribution and API
+ */
+public class LinkedHashSetQueue<T> extends AbstractQueue<T> {
+	private final LinkedHashSet<T> data;
+
+	public LinkedHashSetQueue(LinkedHashSet<T> newData) {
+		this.data = newData;
```

Comment:
A general purpose impl would probably create a defensive copy of newData

Reply:
agreed